--- ../src-base/minecraft/net/minecraft/block/BlockPortal.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPortal.java
@@ -14,6 +14,7 @@
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityList;
 import net.minecraft.entity.monster.EntityPigZombie;
+import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.init.Blocks;
 import net.minecraft.init.SoundEvents;
 import net.minecraft.item.ItemMonsterPlacer;
@@ -74,11 +75,17 @@
 
             if (i > 0 && !p_180650_1_.func_180495_p(blockpos.func_177984_a()).func_185915_l())
             {
-                Entity entity = ItemMonsterPlacer.func_77840_a(p_180650_1_, EntityList.func_191306_a(EntityPigZombie.class), (double)blockpos.func_177958_n() + 0.5D, (double)blockpos.func_177956_o() + 1.1D, (double)blockpos.func_177952_p() + 0.5D);
-
-                if (entity != null)
-                {
-                    entity.field_71088_bW = entity.func_82147_ab();
+                Entity entity = ItemMonsterPlacer.func_77840_a(p_180650_1_, EntityList.func_191306_a(EntityPigZombie.class), (double) blockpos.func_177958_n() + 0.5D, (double) blockpos.func_177956_o() + 1.1D, (double) blockpos.func_177952_p() + 0.5D);
+                if (entity != null) {
+                    // Only spawn the pigman if a player is within the normal range.
+                    // Fixes issue with chunk loaders
+                    for (EntityPlayer player : p_180650_1_.field_73010_i) {
+                        if (player.func_70068_e(entity) < 16384) {
+                            entity.field_71088_bW = entity.func_82147_ab();
+                            return;
+                        }
+                    }
+                    p_180650_1_.func_72973_f(entity);
                 }
             }
         }
@@ -111,7 +118,7 @@
     {
         BlockPortal.Size blockportal$size = new BlockPortal.Size(p_176548_1_, p_176548_2_, EnumFacing.Axis.X);
 
-        if (blockportal$size.func_150860_b() && blockportal$size.field_150864_e == 0)
+        if (blockportal$size.func_150860_b() && blockportal$size.field_150864_e == 0 && !net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(p_176548_1_, p_176548_2_, blockportal$size))
         {
             blockportal$size.func_150859_c();
             return true;
@@ -120,7 +127,7 @@
         {
             BlockPortal.Size blockportal$size1 = new BlockPortal.Size(p_176548_1_, p_176548_2_, EnumFacing.Axis.Z);
 
-            if (blockportal$size1.func_150860_b() && blockportal$size1.field_150864_e == 0)
+            if (blockportal$size1.func_150860_b() && blockportal$size1.field_150864_e == 0 && !net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(p_176548_1_, p_176548_2_, blockportal$size1))
             {
                 blockportal$size1.func_150859_c();
                 return true;
